### PR TITLE
feat: add Gemini sample app: "Fix My Car" [WIP]

### PR DESCRIPTION
**Pull request by askmeegs**
# Description

Fix My Car is a generative AI sample application using Gemini, Vertex AI Search, Cloud SQL, and Streamlit/Java on GKE. It is an auto manual helper chatbot. 

There are two separate flavors of the app, packaged separately. One uses Vertex AI Search for grounding, and the other uses Cloud SQL with pgvector. 

Note: This app can use any PDF owner's manuals for grounding. For this sample version, we use a Gemini-generated manual for a mock car model called "Cymbal Starlight." 

### checklist 

- [x] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [x] You are listed as the author in your notebook or README file.
  - [ ] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [ ] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [ ] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [ ] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: gemini/sample-apps/fixmycar/vertex-ai-search/LICENSE
Comment (holtskinner) [inline gemini/sample-apps/fixmycar/vertex-ai-search/LICENSE:1]: The whole repo has these LICENSE files, so they might not be needed here.

File: gemini/sample-apps/fixmycar/vertex-ai-search/backend/.mvn/wrapper/maven-wrapper.jar
Comment (holtskinner) [inline gemini/sample-apps/fixmycar/vertex-ai-search/backend/.mvn/wrapper/maven-wrapper.jar:1]: Are these .mvn files needed?

File: gemini/sample-apps/fixmycar/vertex-ai-search/backend/src/main/java/com/cpet/fixmycarbackend/ChatMessage.java
Comment (holtskinner) [inline gemini/sample-apps/fixmycar/vertex-ai-search/backend/src/main/java/com/cpet/fixmycarbackend/ChatMessage.java:1]: This might be more trouble than its worth, but I wonder if it would be possible to combine the Cloud SQL and Vertex AI Search apps into a single app, with an option to select which backend gets used?

Seems like a large amount of these apps are the same.

**Review by holtskinner — CHANGES_REQUESTED** [review]
Looks good overall, I mostly have questions about the general app structure & included files (I'm not super experienced in Java Apps, so I don't know how much of these binary/maven files are needed for a repo and which should be gitignored

**Comment by holtskinner** [comment]
Also, don't worry _too_ much about the linter errors. Fix the ones that can be done quickly/easily or that represent a reasonable issue.

File: gemini/sample-apps/fixmycar/cloud-sql/manuals/GenerateMockAutoManual.ipynb [hunk 1]
```diff
@@ -0,0 +1,311 @@
+{
```
Comment (holtskinner) [inline gemini/sample-apps/fixmycar/cloud-sql/manuals/GenerateMockAutoManual.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    ! pip install "google-cloud-aiplatform&gt;=1.38"
</pre><p>This file is duplicated. I'd recommend only keeping the one in the Manuals Directory. Just make sure there's nothing in the cloudsql directory one that needs to be transferred over.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb [hunk 2]
```diff
@@ -0,0 +1,425 @@
+{
```
Comment (holtskinner) [inline gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #1.    print("hello world")
</pre><p>Remove this</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #2.    PROJECT_ID = 'PROJECT-ID'
</pre><p>Also create a constant for LOCATION/region and use it in the vertexai.init function.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #2.    vertexai.init(project=PROJECT_ID)
</pre><p>Move the init to right below the imports.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #3.    print(text_embedding("hello world!"))
</pre><p>Not sure this line is needed</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #2.    d = {}
</pre><p>Maybe make this a more specific variable name?</p><p> </p><p>Or you could also load this directly into the data frame</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #1.    import pandas as pd
</pre><p>Add import to top of notebook with other imports</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #2.    df = df[['name', 'embedding']]
</pre><p>Why is this needed? It doesn't seem like anything changed</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

File: gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb [hunk 3]
```diff
@@ -0,0 +1,311 @@
+{
```
Comment (holtskinner) [inline gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: <p>Remove the installation/setup cell output</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: <p>Remove your project ID, create a CONSTANT for it before running this line</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #2.    vertexai.init(project="cpet-sandbox", location="us-central1")
</pre><p>Use project/location constants.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: <p>Remove this cell</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: <p>material is as much detail as possible → material in as much detail as possible</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: <p>Why are there two different values for topics?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: <p>Does this do anything?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (holtskinner) [inline gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #1.    finaltext = ""
</pre><p>for this you could do <code>final_text = "".join(manual)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/generative-ai/pull/458/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

File: gemini/sample-apps/fixmycar/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+**[Click here to get started.](cloud-sql/)**
+
+## Learn more
+
+To learn more about Retrieval-Augmented Generation on Google Cloud, see:
+
+- [Getting Started with Gen AI on Google Cloud - Grounding](https://cloud.google.com/docs/ai-ml/generative-ai#grounding)
+- [Databases Retrieval App (SFO Airport Helper)](https://github.com/GoogleCloudPlatform/genai-databases-retrieval-app) (Cloud SQL or AlloyDB)
+- [RAG Reference Architecture on Google Cloud with Vertex AI and AlloyDB](https://cloud.google.com/architecture/rag-capable-gen-ai-app-using-vertex-ai) + [Codelab](https://codelabs.developers.google.com/codelabs/genai-db-retrieval-app)
+- [Blog post: Your RAGs powered by Google Search technology, part 1](https://cloud.google.com/blog/products/ai-machine-learning/rags-powered-by-google-search-technology-part-1)
```
[the remark this answers is not in the document]
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/README.md:43]: Done.

File: gemini/sample-apps/fixmycar/cloud-sql/backend/mvnw
[the remark this answers is not in the document]
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/backend/mvnw:1]: The mvnw files are for Maven Wrapper. this is the build mechanism for the java spring backend. Committing them was intentional, they allow someone to clone the repo and build the java app without installing Maven. 
https://stackoverflow.com/questions/47240546/should-the-mvnw-files-be-added-to-the-repository

File: gemini/sample-apps/fixmycar/cloud-sql/frontend/bin/Activate.ps1
[the remark this answers is not in the document]
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/frontend/bin/Activate.ps1:1]: removed!

File: gemini/sample-apps/fixmycar/cloud-sql/frontend/share/jupyter/nbextensions/pydeck/index.js
[the remark this answers is not in the document]
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/frontend/share/jupyter/nbextensions/pydeck/index.js:1]: removed!

File: gemini/sample-apps/fixmycar/cloud-sql/manuals/cymbal-starlight-2024.pdf
[the remark this answers is not in the document]
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/manuals/cymbal-starlight-2024.pdf:1]: Sure, uploaded.

File: gemini/sample-apps/fixmycar/vertex-ai-search/backend/.mvn/wrapper/maven-wrapper.jar
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/vertex-ai-search/backend/.mvn/wrapper/maven-wrapper.jar:1]: Removed, but kept the `mvnw` files.

File: gemini/sample-apps/fixmycar/vertex-ai-search/backend/src/main/java/com/cpet/fixmycarbackend/ChatMessage.java
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/vertex-ai-search/backend/src/main/java/com/cpet/fixmycarbackend/ChatMessage.java:1]: +1 that a large % of the apps are the same — I did attempt to combine the backends into one configurable backend. the TLDR is that due to the way Spring JPA (the Cloud SQL flavor) works, there is no clear-cut way to toggle it on and off. So when a user has Vertex AI Search enabled, and doesn't configure a Cloud SQL endpoint, the app won't start. 

Full commit history/saga -- 
https://github.com/cloud-platform-insights/fixmycar/commits/configurable/

I am also not a java expert, so longer term I'd like to ask around and see if I was just missing something during the implementation. It would certainly be cleaner to combine the flavors. another alternative is to only publish one flavor (Vertex AI Search) and only add Cloud SQL later. Open to thoughts.

File: gemini/sample-apps/fixmycar/cloud-sql/manuals/GenerateMockAutoManual.ipynb (hunk 1 above)
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/manuals/GenerateMockAutoManual.ipynb:1]: Removed.

File: gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb (hunk 2 above)
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: Removed.

Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: Removed.

Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: changed `d` to `embeddings_dict`

Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: Done

Comment (askmeegs) [reply gemini/sample-apps/fixmycar/cloud-sql/pgvector/generate-embeddings.ipynb]: It's a column rename, from `text` to `name`, fixed so that we're calling it `text` in the first place.

File: gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb (hunk 3 above)
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: updated.

Comment (askmeegs) [reply gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: Fixed, was doing follow up prompt engineering on a subset of topics

Comment (askmeegs) [reply gemini/sample-apps/fixmycar/manuals/GenerateMockAutoManual.ipynb]: thanks, fixed! I was originally doing some pre-pending with chapter titles and formatting, hence the loop, but ended up removing.

File: gemini/sample-apps/fixmycar/vertex-ai-search/backend/src/main/java/com/cpet/fixmycarbackend/ChatMessage.java
Comment (holtskinner) [reply gemini/sample-apps/fixmycar/vertex-ai-search/backend/src/main/java/com/cpet/fixmycarbackend/ChatMessage.java:1]: Ok, that makes sense. That's weird for how the Cloud SQL must be set up. For a smaller fix, would it be possible to consolidate the front end code into a separate directory and have both backends deploy with the same basic front end code?

File: workspace.code-workspace
Comment (holtskinner) [inline workspace.code-workspace:1]: This should probably be added to gitignore

File: gemini/sample-apps/fixmycar/vertex-ai-search/backend/src/main/java/com/cpet/fixmycarbackend/ChatMessage.java
Comment (askmeegs) [reply gemini/sample-apps/fixmycar/vertex-ai-search/backend/src/main/java/com/cpet/fixmycarbackend/ChatMessage.java:1]: Hi Holt, medium term, yes, I would like to consolidate the frontends -- but because of the way the deploy scripts are set up, it is easier to have basically two "root" directories w/ everything needed under those directories. 
eg. 
https://github.com/askmeegs/generative-ai/blob/fixmycar-mock-manual/gemini/sample-apps/fixmycar/cloud-sql/dockerpush.sh 
vs.
https://github.com/askmeegs/generative-ai/blob/fixmycar-mock-manual/gemini/sample-apps/fixmycar/vertex-ai-search/dockerpush.sh

Between now and Cloud Next, would it be possible to keep it as is? I can add a to-do to clean up the frontends in the weeks after.

Comment (holtskinner) [reply gemini/sample-apps/fixmycar/vertex-ai-search/backend/src/main/java/com/cpet/fixmycarbackend/ChatMessage.java:1]: Yeah, that's okay with me
